### PR TITLE
Implement OverlappingFieldsCanBeMerged

### PR DESCRIPTION
**Pull request by jhgg**
* Fix errors in `type_map_reducer`.
* Implement `expect_{passes,fails}_rule_with_schema`
* Add support to opt out of argument sorting in `expect_invalid`.
* Implement `PairSet` and `DefaultOrderedDict`.
* Warn if schema does not contain unique named types.
* Implement `is_same_type` onto `GraphQLType`, `GraphQLList` and
`GraphQLNonNull`.

### CODE REVIEW

File: graphql/core/validation/rules.py [hunk 1]
```diff
@@ -731,8 +737,233 @@ def var_type_allowed_for_type(cls, var_type, expected_type):
...
+            if not cls.same_value(argument1.value, argument2.value):
+                return False
+
+        return True
+
+    @classmethod
+    def same_directives(cls, directives1, directives2):
+        # Check to see if they are empty directives or nones. If they are, we can
+        # bail out early.
+        if not (directives1 and directives2):
```
Comment (dittos) [inline graphql/core/validation/rules.py]: I think this should be `not directives1 and not directives2`. (May add test case for this... One with directive and without directive.)

File: graphql/core/validation/rules.py [hunk 2]
```diff
@@ -731,8 +737,233 @@ def var_type_allowed_for_type(cls, var_type, expected_type):
...
+
+    @staticmethod
+    def same_value(value1, value2):
+        return (not value1 and not value2) or print_ast(value1) == print_ast(value2)
+
+    @classmethod
+    def same_arguments(cls, arguments1, arguments2):
+        # Check to see if they are empty arguments or nones. If they are, we can
+        # bail out early.
+        if not (arguments1 and arguments2):
```
Comment (dittos) [inline graphql/core/validation/rules.py]: Same here. `someField` and `someField(arg: someArg)` differ

File: graphql/core/validation/rules.py (hunk 1 above)
Comment (jhgg) [inline graphql/core/validation/rules.py]:
```>>> directive1 = False
>>> directive2 = False
>>> not (directive1 and directive2)
True
>>> directive2 = True
>>> not (directive1 and directive2)
True
>>> directive1 = True
>>> not (directive1 and directive2)
False
>>> directive2 = False
>>> not (directive1 and directive2)
True```

The logic works out ;P

**Comment by dittos** [comment]
Phew!

**Pull request merged**